### PR TITLE
Variable name case standards 

### PR DESCRIPTION
**Pull request by rstoneback**
# Description
pysat began life with an assumption that all variable names would be lower case. Earlier versions imposed this, I believe. Work on ICON has forced some changes to the code to support names with non-lowercase letters, seen most explicitly in the Meta object. Within meta, the first instance of a string sets the case, any later variations are treated as the original variable. This is particularly good when one person thinks an attribute should be 'units', and another 'Units', or another 'UNITS'. Meta treats them all the same. This significantly helps Meta maintain compatibility with a larger number of file and data standards. Particularly in the context of community development.

For ICON, many variables have mixed case.  The IVM software previously made everything lower case while in-house, then made everything upper case for the output file. New standards for level-2 files have long filenames, underscores, and upper and lower letters. 

What should pysat standards be?
1) Accept variable names as is. Has the problem that Meta is case-insensitive. So the data could have two variables, 'mlt', and 'MLT', with only one Metadata entry.
2) Force all data variables to upper case. No.
3) Force all data variable names to lower case. Fancier name stored in Metadata so output file names are supported. We would need to make some changes within pysat to support this.
   - Update to_netCDF to use metadata names. Write new tests around this.
   - Update from_netCDF to use lower case names on the data side, preserve case on Meta side.
   - Modify the setitem method to ensure variable names are always lower case
   - Create a Instrument.data attribute that checks variables names any time someone sets their own data set within an Instrument object, e.g. inst.data = new_data , to make sure all variable names are lower case.

Right now we are pretty much 1, though there may be some residual code floating around. 

This pull request is intended to start a conversation. (See update below - I now propose a more limited pull)

Addresses # 260

Adds new file attributes when using to_netCDF4 so that compatibility with ICON file standards is maintained.

Makes first step to supporting option 3 above. Uses the variable names within the Meta object when naming variables within the netCDF file.
 - Update: Functionality has been implemented as an option when writing a file. I propose remaining at option 1 as it involves the least user confusion. Variable names are taken as given by a user or data set. If a user decides, they can now exploit more of pysat's functionality with the file write option here.

## Type of change

Please delete options that are not relevant.

- New Feature (updated since original pull)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. Please also list any relevant details for
your test configuration

- Created files using the ICON IVM software. 
- Added tests that use new feature when writing files. I create test Instruments that have a difference in case between data and metadata and ensure the correct behavior is obtained when the feature is and isn't engaged.
- Added tests to check for error conditions.
- Passes normal TravisCI tests.

**Test Configuration**:
TravisCI

# Checklist:

- [X] Make sure you are merging into the ``develop`` (not ``master``) branch
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

File: pysat/_instrument.py [hunk 1]
```diff
@@ -2026,7 +2026,7 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
         # to the main input Instrument will be written to the netCDF4
         base_instrument = Instrument() if base_instrument is None \
             else base_instrument
-
+        print('Latest')
```
Comment (rstoneback) [inline pysat/_instrument.py]: I'll get this removed.

**Comment by rstoneback** [comment]
pinging @aburrell and @jklenzing for comments. FYI Jeff won't be available until the 22nd.

**Comment by aburrell** [comment]
I thought you had done this ages ago.

**Comment by aburrell** [comment]
And that was a slip of the fingers...

**Comment by rstoneback** [comment]
I mostly did do this ages ago. Except when I went to make an updated icon file the other day the variables didn’t quite come out correct. I should’ve done a better job :) Looks like I missed some areas.

On Aug 15, 2019, at 2:59 PM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


And that was a slip of the fingers...

**Comment by rstoneback** [comment]
I've had a chance to further review the ICON code as well as the pysat code.

The issue that first popped up was mostly on ICON's side. I was able to get proper support going for ICON files in pysat as far back as v1.2. Didn't try earlier.

As to the three options, outlined above:
1) I think we should remain here. In an operational context the biggest issue with a loss of metadata is when a file is written. We can add a check at file writing to see if there are multiple names in data with different case and print a warning, or error, then.

2) No.

3) After further consideration, I think we should avoid this option. While there are some benefits it can also be confusing, on a number of levels. If folks want to have lower case names in data, as well as mixed case in Meta, it is easy enough to do on the outside of pysat. Thus I think we should drop this option.

In other words, though I brought this issue up, I'd also like to let it go. I'm going to modify this pull a little bit to keep printing the MetaData variable names as an option, as well as add a name check at the beginning.

**Comment by rstoneback** [comment]
Summary as committed:
 - Added new file attributes to be consistent with ICON
 - Added option to use names from MetaData
 - pysat standards otherwise remain the same
 - Added a check when writing a file with non-uniuque variable names. An error is now raised warning that metadata has been lost. 
 - Added tests for new feature
 - Added test for new Error
 - modified the changelog

**Comment by rstoneback** [comment]
This should be ready for review. Tests just started as a I type this though my most recent changes are limited to docstrings. Previous commits passed.

**Comment by rstoneback** [comment]
Test failures are related to connection failures on Travis CI's side. Unrelated to code updates here.

File: pysat/_instrument.py [hunk 2]
```diff
@@ -2257,15 +2287,15 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
                                 data, coltype, _ = \
                                     self._get_data_info(idx, file_format)
                                 cdfkey = \
-                                    out_data.createVariable(key+'_'+col,
+                                    out_data.createVariable(case_key + '_' + col,
```
Comment (jklenzing) [inline pysat/_instrument.py]: pep8! :D

File: pysat/tests/test_utils.py [hunk 3]
```diff
@@ -146,6 +147,51 @@ def test_basic_write_and_read_netcdf4_default_format(self):
...
+        # write file
+        self.testInst.to_netcdf4(outfile, preserve_meta_case=True)
+        # load file
+        loaded_inst, meta = pysat.utils.load_netcdf4(outfile)
+
+        # check that names are upper case when written
+        assert(np.all(sorted(self.testInst.data.columns) == sorted(loaded_inst.columns)))
+
+    @raises(Exception)
+    def test_write_netcdf4_doubeled_variable_names(self):
```
Comment (jklenzing) [inline pysat/tests/test_utils.py]: sp: doubled

Should this be duplicate?

File: pysat/tests/test_utils.py [hunk 4]
```diff
@@ -146,6 +147,51 @@ def test_basic_write_and_read_netcdf4_default_format(self):
...
+        loaded_inst = loaded_inst.reindex(sorted(loaded_inst.columns), axis=1)
+
+        # check that names are lower case when written
+        assert(np.all(original == loaded_inst.columns))
+        
+        for key in self.testInst.data.columns:
+            print('Testing Data Equality to filesystem and back ', key)
+            assert(np.all(self.testInst[key] == loaded_inst[key.lower()]))
+
+        # modify meteadata names in data
```
Comment (jklenzing) [inline pysat/tests/test_utils.py]: sp: metadata

File: pysat/_instrument.py [hunk 5]
```diff
@@ -2415,28 +2445,44 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
...
             adict['Date_End'] = \
                 pysat.datetime.strftime(self.index[-1],
                                         '%a, %d %b %Y,  ' +
-                                        '%Y-%m-%dT%H:%M:%S.%f UTC')
+                                        '%Y-%m-%dT%H:%M:%S.%f')
+            adict['Date_End'] = adict['Date_End'][:-3] + ' UTC'
+            
+            adict['Date_Start'] = \
+                pysat.datetime.strftime(self.index[0],
+                                    '%a, %d %b %Y,  ' +
```
Comment (jklenzing) [inline pysat/_instrument.py]: tab to parentheses

**Comment by rstoneback** [comment]
Thanks for the review Jeff. I appreciate it. 

Cheers,
Russell

**Comment by rstoneback** [comment]
Thanks again Jeff. Minor corrections are in. I'll merge into develop after the tests have a chance to do their thing.
